File: statusline.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strings"
)

type StatusInput struct {
	Model struct {
		DisplayName string `json:"display_name"`
	} `json:"model"`
	Workspace struct {
		CurrentDir string `json:"current_dir"`
	} `json:"workspace"`
	ContextWindow struct {
		UsedPercentage *float64 `json:"used_percentage"`
	} `json:"context_window"`
}

func statusline() error {
	var input StatusInput
	if err := json.NewDecoder(os.Stdin).Decode(&input); err != nil {
		return err
	}

	cwd := input.Workspace.CurrentDir
	if home, _ := os.UserHomeDir(); home != "" && strings.HasPrefix(cwd, home) {
		cwd = "~" + cwd[len(home):]
	}

	out := fmt.Sprintf("🤖 %s | 📁 %s", input.Model.DisplayName, cwd)

	if branch := getGitBranch(input.Workspace.CurrentDir); branch != "" {
		out += fmt.Sprintf(" | 🌿 %s", branch)
	}
	if input.ContextWindow.UsedPercentage != nil {
		out += fmt.Sprintf(" | 💭 %.0f%%", *input.ContextWindow.UsedPercentage)
	}

	fmt.Println(out)
	return nil
}

func getGitBranch(dir string) string {
	out, err := exec.Command("git", "-C", dir, "branch", "--show-current").Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

File: src/statusline.ts
import { homedir } from "os";
import { getBranch } from "./git";

interface StatusInput {
  model?: {
    display_name?: string;
  };
  workspace?: {
    current_dir?: string;
  };
  cost?: {
    total_cost_usd?: number;
  };
  context_window?: {
    used_percentage?: number;
  };
}

export async function statusline(): Promise<void> {
  const input = await Bun.stdin.text();
  if (!input.trim()) {
    return;
  }

  let data: StatusInput;
  try {
    data = JSON.parse(input);
  } catch {
    console.error("Failed to parse JSON input");
    process.exit(1);
  }

  const home = homedir();
  const branch = await getBranch(data.workspace?.current_dir);
  const output = formatStatusOutput(
    data.model?.display_name,
    replaceTilde(data.workspace?.current_dir, home),
    branch,
    data.cost?.total_cost_usd,
    data.context_window?.used_percentage
  );
  console.log(output);
}

export function replaceTilde(path: string | undefined, home: string): string {
  if (!path) return "";
  if (home && path.startsWith(home)) {
    return "~" + path.slice(home.length);
  }
  return path;
}

export function formatCost(cost: number): string {
  if (cost < 0.01) {
    return `$${cost.toFixed(4)}`;
  }
  return `$${cost.toFixed(2)}`;
}

export function formatStatusOutput(
  model: string | undefined,
  cwd: string,
  branch: string,
  costUsd: number | undefined,
  usedPercentage: number | undefined
): string {
  const parts = [`🤖 ${model || ""}`, `📁 ${cwd}`];
  if (branch) {
    parts.push(`🌿 ${branch}`);
  }
  if (costUsd !== undefined) {
    parts.push(`💰 ${formatCost(costUsd)}`);
  }
  if (usedPercentage !== undefined) {
    parts.push(`💭 ${Math.round(usedPercentage)}%`);
  }
  return parts.join(" | ");
}
